nkms/characters.py

```diff
@@ -372,33 +400,64 @@ def __init__(self, *args, **kwargs):
         from nkms.policy.models import WorkOrderHistory  # Need a bigger strategy to avoid circulars.
         self._saved_work_orders = WorkOrderHistory()
 
-    def follow_treasure_map(self, hrac):
-        for ursula_interface_id in self.treasure_maps[hrac]:
-            if ursula_interface_id in self.known_nodes:
-                # If we already know about this Ursula,
-                # we needn't learn about it again.
-                continue
-
-            # TODO: perform this part concurrently.
-            value = self.server.get_now(ursula_interface_id)
-
-            # TODO: Make this much prettier
-            header, signature, ursula_pubkey_sig, _hrac, (
-                port, interface, ttl) = dht_value_splitter(value, msgpack_remainder=True)
-
-            if header != constants.BYTESTRING_IS_URSULA_IFACE_INFO:
-                raise TypeError("Unknown DHT value.  How did this get on the network?")
-
-            # TODO: If we're going to implement TTL, it will be here.
-            self.known_nodes[ursula_interface_id] = \
-                Ursula.as_discovered_on_network(
-                    dht_port=port,
-                    dht_interface=interface,
-                    powers_and_keys=({SigningPower: ursula_pubkey_sig})
-                )
-
-    def get_treasure_map(self, policy, networky_stuff, using_dht=False):
-        map_id = policy.treasure_map_dht_key()
+    def follow_treasure_map(self, hrac, using_dht=False):
+
+        treasure_map = self.treasure_maps[hrac]
+        number_of_known_treasure_ursulas = 0
+        if not using_dht:
+            for ursula_interface_id in treasure_map:
+                pubkey = UmbralPublicKey.from_bytes(ursula_interface_id)
+                if pubkey in self.known_nodes:
+                    number_of_known_treasure_ursulas += 1
+
+            newly_discovered_nodes = {}
+            nodes_to_check = iter(self.known_nodes.values())
+
+            while number_of_known_treasure_ursulas < treasure_map.m:
```

Comment:
This might allow Alice to collude with Ursula nodes to have Bob traverse the map in an order designated by her. Whether or not this can be done maliciously is something I haven't thought about, but certainly could have some implications somewhere.

For example, long ago in the revolutionary war when the first US spy rings were made, knowing the order that a spy would travel to deliver/receive/drop information would be enough to correlate who the spy was and would have horrible consequences for the entire spy ring. If Alice can influence the order that Bob traverses the nodes, this might have bad consequences. In practice, the Tor network is vulnerable to similar attacks that involve timing.

Perhaps we randomize the map before we traverse it? I know that dictionaries are already randomized to an extent, but might it be enough to thwart something here.

Reply:
As I said earlier in person, I'm not convinced there's any possible attack based on this, but we might as well do it.  It also has the possible benefit of relieving the "first in line" Ursulas in the case of a high-throughput `DataSource`-=>`Bob` pair.

Reply:
See #245